sql/stats: TestCreateStatsProgress failed

sql/stats.TestCreateStatsProgress [failed](https://tanzanite.cluster.engflow.com/invocations/default/b21e54e3-20dd-4792-abe1-9ac4d38e9269?testReportRun=3&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0YXRzOnN0YXRzX3Rlc3Q=) on master @ [0ea8b622c5354a891cfb867d8d872a0746ba77b3](https://github.com/cockroachdb/cockroach/commits/0ea8b622c5354a891cfb867d8d872a0746ba77b3):

Fatal error:

```
panic: test timed out after 16m37s
running tests:
	TestCreateStatsProgress (16m33s)

```
Stack: 

```
goroutine 8275115 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12({0x6900338, 0xc00c2d15c0}, {0x5378220, 0xc012b15e60}, 0xc00b628300, 0xc014a17cc8)
* 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97 +0x4b2
* google.golang.org/grpc.getChainUnaryHandler.func1({0x6900338, 0xc00c2d15c0}, {0x5378220, 0xc012b15e60})
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0xb2
* github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3({0x6900338, 0xc00c2d15c0}, {0x5378220, 0xc012b15e60}, 0xc00b628300?, 0xc011654880)
* 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169 +0x76
* google.golang.org/grpc.getChainUnaryHandler.func1({0x6900338, 0xc00c2d15c0}, {0x5378220, 0xc012b15e60})
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0xb2
* github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({0xc007856000?, {{0x44a405?}, {0x69444a0?, 0xc003e00240?}}}, {0x6900338, 0xc00c2d0f30}, {0x5378220, 0xc012b15e60}, 0xc00b628300, 0xc011654740)
* 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95 +0x1e4
* google.golang.org/grpc.getChainUnaryHandler.func1({0x6900338, 0xc00c2d0f30}, {0x5378220, 0xc012b15e60})
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0xb2
* github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1({0x6900338?, 0xc00c2d0f30?})
* 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136 +0x36
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0031e4e10, {0x6900338, 0xc00c2d0f30}, {0xc011654640?, 0xc0086bd978?}, 0xc0086bd970)
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337 +0xcb
* github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1({0x6900338?, 0xc00c2d0f30?}, {0x5378220?, 0xc012b15e60?}, 0xc00b628300?, 0xc014a17cc8?)
* 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134 +0x8f
* google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x6900338, 0xc00c2d0f30}, {0x5378220, 0xc012b15e60}, 0xc00b628300, 0x70?)
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0x85
* github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_Batch_Handler({0x5359ec0, 0xc00af9dc08}, {0x6900338, 0xc00c2d0f30}, 0xc00c64c150, 0xc00ad02520)
* 	github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/k8-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:10635 +0x143
* google.golang.org/grpc.(*Server).processUnaryRPC(0xc0059e2f00, {0x694e180, 0xc00f09fba0}, 0xc012b158c0, 0xc007d45650, 0x9b86520, 0x0)
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1336 +0xd16
* google.golang.org/grpc.(*Server).handleStream(0xc0059e2f00, {0x694e180, 0xc00f09fba0}, 0xc012b158c0, 0x0)
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1704 +0x9da
* google.golang.org/grpc.(*Server).serveStreams.func1.2()
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965 +0x8d
* created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 738906
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:963 +0x226
* 
* goroutine 472221 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:500 +0x14d
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 460826
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:471 +0x4f
* 
* goroutine 1355415 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:253 +0xcb
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 471540
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:248 +0x58
* 
* goroutine 1111710 [chan send, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/sql/stats_test.TestCreateStatsProgress.func6()
* 	github.com/cockroachdb/cockroach/pkg/sql/stats_test/pkg/sql/stats/create_stats_job_test.go:504 +0x6d
* created by github.com/cockroachdb/cockroach/pkg/sql/stats_test.TestCreateStatsProgress in goroutine 460826
* 	github.com/cockroachdb/cockroach/pkg/sql/stats_test/pkg/sql/stats/create_stats_job_test.go:502 +0xa8c
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCreateStatsProgress.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39049

--- a/pkg/sql/stats/create_stats_job_test.go
+++ b/pkg/sql/stats/create_stats_job_test.go
@@ -31,6 +31,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/jobutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
@@ -389,6 +390,9 @@ func TestCreateStatsProgress(t *testing.T) {
 	}(rowexec.SamplerProgressInterval)
 	rowexec.SamplerProgressInterval = 10
 
+	skip.UnderRace(t, "the test is too sensitive to overload")
+	skip.UnderDeadlock(t, "the test is too sensitive to overload")
+
 	getLastCreateStatsJobID := func(t testing.TB, db *sqlutils.SQLRunner) jobspb.JobID {
 		var jobID jobspb.JobID
 		db.QueryRow(t, "SELECT id FROM system.jobs WHERE status = 'running' AND "+
@@ -397,24 +401,29 @@ func TestCreateStatsProgress(t *testing.T) {
 	}
 
 	var allowRequest chan struct{}
-	var serverArgs base.TestServerArgs
+	var allowRequestClosed bool
+	// Make sure that we unblock the test server in all scenarios with test
+	// failures.
+	defer func() {
+		if !allowRequestClosed {
+			close(allowRequest)
+		}
+	}()
 	filter, setTableID := createStatsRequestFilter(&allowRequest)
-	params := base.TestClusterArgs{ServerArgs: serverArgs}
-	params.ServerArgs.Knobs.Store = &kvserver.StoreTestingKnobs{
+	var params base.TestServerArgs
+	params.Knobs.Store = &kvserver.StoreTestingKnobs{
 		TestingRequestFilter: filter,
 	}
-	params.ServerArgs.Knobs.DistSQL = &execinfra.TestingKnobs{
+	params.Knobs.DistSQL = &execinfra.TestingKnobs{
 		// Force the stats job to iterate through the input rows instead of reading
 		// them all at once.
 		TableReaderBatchBytesLimit: 100,
 	}
 
 	ctx := context.Background()
-	const nodes = 1
-	tc := testcluster.StartTestCluster(t, nodes, params)
-	defer tc.Stopper().Stop(ctx)
-	s := tc.ApplicationLayer(0)
-	conn := s.SQLConn(t)
+	srv, conn, _ := serverutils.StartServer(t, params)
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
 	sqlDB := sqlutils.MakeSQLRunner(conn)
 
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false`)
@@ -476,6 +485,7 @@ func TestCreateStatsProgress(t *testing.T) {
 	// Allow the job to complete and verify that the client didn't see anything
 	// amiss.
 	close(allowRequest)
+	allowRequestClosed = true
 	if err := <-errCh; err != nil {
 		t.Fatalf("create stats job should have completed: %s", err)
 	}
@@ -490,15 +500,12 @@ func TestCreateStatsProgress(t *testing.T) {
 	}
 
 	// Invalidate the stats cache so that we can be sure to get the latest stats.
-	var tableID descpb.ID
-	sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name = 't'`).Scan(&tableID)
-	s.ExecutorConfig().(sql.ExecutorConfig).TableStatsCache.InvalidateTableStats(
-		ctx, tableID,
-	)
+	s.ExecutorConfig().(sql.ExecutorConfig).TableStatsCache.InvalidateTableStats(ctx, tID)
 
 	// Start another CREATE STATISTICS run and wait until it has scanned part of
 	// the table.
 	allowRequest = make(chan struct{})
+	allowRequestClosed = false
 	go func() {
 		_, err := conn.Exec(query)
 		errCh <- err
@@ -532,6 +539,7 @@ func TestCreateStatsProgress(t *testing.T) {
 	// Allow the job to complete and verify that the client didn't see anything
 	// amiss.
 	close(allowRequest)
+	allowRequestClosed = true
 	if err := <-errCh; err != nil {
 		t.Fatalf("create stats job should have completed: %s", err)
 	}

